tests/text_input_v3_with_input_method_v2.cpp

```diff
@@ -213,3 +214,83 @@ TEST_F(TextInputV3WithInputMethodV2Test, input_method_can_send_preedit)
     input_client.roundtrip();
     app_client.roundtrip();
 }
+
+TEST_F(TextInputV3WithInputMethodV2Test, text_input_does_not_enter_non_grabbing_popup)
+{
+    auto parent_surface = std::make_unique<wlcs::Surface>(app_client);
+    EXPECT_CALL(text_input, enter(parent_surface->wl_surface()));
+    auto parent_xdg_surface = std::make_shared<wlcs::XdgSurfaceStable>(app_client, *parent_surface);
+    auto parent_xdg_toplevel = std::make_shared<wlcs::XdgToplevelStable>(*parent_xdg_surface);
+    parent_surface->attach_visible_buffer(20, 20);
+    app_client.roundtrip();
+    Mock::VerifyAndClearExpectations(&text_input);
+    auto child_surface = std::make_unique<wlcs::Surface>(app_client);
+    EXPECT_CALL(text_input, leave(_)).Times(0);
+    EXPECT_CALL(text_input, enter(_)).Times(0);
+    auto child_xdg_surface = std::make_shared<wlcs::XdgSurfaceStable>(app_client, *child_surface);
+    auto child_xdg_toplevel = std::make_shared<wlcs::XdgPopupStable>(
+        *child_xdg_surface,
+        parent_xdg_surface.get(),
+        wlcs::XdgPositionerStable{app_client}.setup_default({20, 20}));
+    child_surface->attach_visible_buffer(20, 20);
+    app_client.roundtrip();
+}
+
+TEST_F(TextInputV3WithInputMethodV2Test, text_input_enters_grabbing_popup)
+{
+    auto parent_surface = std::make_unique<wlcs::Surface>(app_client);
+    EXPECT_CALL(text_input, enter(parent_surface->wl_surface()));
+    auto parent_xdg_surface = std::make_shared<wlcs::XdgSurfaceStable>(app_client, *parent_surface);
+    auto parent_xdg_toplevel = std::make_shared<wlcs::XdgToplevelStable>(*parent_xdg_surface);
+    parent_surface->attach_visible_buffer(20, 20);
+    the_server().move_surface_to(*parent_surface, 0, 0);
+    app_client.roundtrip();
+    Mock::VerifyAndClearExpectations(&text_input);
+
+    // This is needed to get a serial, which will be used later on
+    auto pointer = the_server().create_pointer();
+    pointer.move_to(2, 2);
+    pointer.left_click();
+    app_client.roundtrip();
+
+    auto child_surface = std::make_unique<wlcs::Surface>(app_client);
+    EXPECT_CALL(text_input, leave(parent_surface->wl_surface()));
+    EXPECT_CALL(text_input, enter(child_surface->wl_surface()));
```

Comment:
Non-blocking, but this ordering is a protocol requirement, so we should probably enforce it.